VCF Overhaul

charlesgregory opened the pull request:
Lots of changes went into this.

fixes #96.

### New Features:
- `HeaderRecord` wrapper struct for getting `VCFHeader` records in a structured format along with getters and setters and `VCFHeader` can now accept this type to add to a header.
- ability to remove header lines `removeHeaderLines`.
- Moved functions from `VCFReader` to `VCFHeader` as appropriate.
-  New `InfoField` and `FormatField` wrapper structs to get and transform VCF format and info data into D types (data is copied).
- New functions (`getInfos` and `getFormats`) to return a hashmap of all info and format data using new wrapper structs. 
- New function to remove info fields `removeInfo`.
- New enum replacements for htslib `BCF*` enums.
- lots of new unittests

Though I did introduce a lot of new features, I have tried to leave old functionality behind. (Though with the reorganization and the enums,  the API will change for most of `dhtslib.vcf`).

### Fixes:
- Fixes to `addFormat` and `addInfo` functions.
- Fixes for `VCFHeader` functions
- `VCFReader` now obeys Phobos Range rules, fixes #69
- Unittest coverage for all `dhtslib.vcf` modules is >= 92%, fixes #94

### Concerns:
1. We sometimes report errors through `hts_log` and have the function return, sometimes we report errors through `hts_log` and have the function throw an exception, or sometimes we just throw an exception. We should decide formally on how this should be done across the whole package, #99 probably relevant.
2. Naming in general. I am not sold on my naming of the new enums, I just wanted them to be more informative than the htslib enums.
3. Could use more asserts and checks

jblachly commented on source/dhtslib/vcf/record.d:
```diff
@@ -663,6 +941,44 @@ struct VCFRecord
+        /// unpack
+        if (this.line.max_unpack < UNPACK.INFO) bcf_unpack(this.line, UNPACK.INFO);
+
+        /// copy some data
```
Can this be made idiomatic D as you've done here but without copying?

charlesgregory replied:
Probably.

charlesgregory replied:
resolved

jblachly commented on source/dhtslib/vcf/record.d:
```diff
@@ -663,6 +941,44 @@ struct VCFRecord
+    }
+
+    /// returns a hashmap of format data by field
+    FormatField[string] getFormats()
```
same -- nocopy?

charlesgregory replied:
resolved

jblachly commented on source/dhtslib/vcf/header.d:
```diff
@@ -1,8 +1,260 @@
+
+
+/// Struct for easy setting and getting of hrec values for VCFheader
+struct HeaderRecord
```
I made this comment elsewhere, but is it possible to write these structs so that they introspect the existing htslib-formatted records in memory (perhaps needing to unpack them, on as needed basis just as htslib does), so that we avoid data copies?

My major concern is that, while very ergonomic, copying headers, copying records, copying info fields into D structs for every row of a large file wil incur massive speed penalty -- perhaps I'm wrong and IO is the limiting factor, but I think you should run some benchmarks on this branch before we accept the PR (or before we try to rewrite it which I acknolwedge would be a big effort)

jblachly replied:
Now, header records are the least offensive since there are a small-ish fixed number per file. My major concern is with row records, and field records (info, format)

jblachly replied:
Can you expoud upon how `HeaderRecord` is used?

charlesgregory replied:
> Now, header records are the least offensive since there are a small-ish fixed number per file. My major concern is with row records, and field records (info, format)

I agree I wasn't bothered by this behavior for the header records and this is why it is the most ergonomic. I think the `InfoField` and `FormatField` structs should probably be changed to avoid the copy.

> Can you expoud upon how HeaderRecord is used?

I thought is would be a nice feature to have as for instance in `recontig` we do a lot of header manipulation during conversion. This involves pulling different parts of the existing header and altering it. I wanted a way I could piece by piece build a header record without resorting to string manipulation. Of course, you should still be able to just specify a header line as a string, but I would like the ability to build it in a more structured way when needed.

jblachly commented on source/dhtslib/vcf/header.d:
```diff
@@ -1,8 +1,260 @@
+
+    /// convert to bcf_hrec_t for use with htslib functions
+    bcf_hrec_t * convert(bcf_hdr_t * hdr)
+    {
```
nice

jblachly commented on source/dhtslib/vcf/header.d:
```diff
@@ -1,8 +1,260 @@
+}
 
 /** Wrapper around bcf_hdr_t
 
```
since we do have SAMHeader this doc comment should be revised

charlesgregory replied:
Agreed

charlesgregory replied:
addressed in
1056615

jblachly commented on source/dhtslib/vcf/header.d:
```diff
@@ -61,4 +318,415 @@ struct VCFHeader
+        // TODO check that key is not INFO, FILTER, FORMAT (or contig?)
+        string line = format("##%s=%s", key, value);
+
+        return bcf_hdr_append(this.hdr, toStringz(line));
```
probably need `bcf_hdr_sync` call here (see `addHeaderLineRaw` ; also check `bcf_hdr_append` to make sure it doesn't call it -- `bcf_hdr_add_sample` doesn't, see my github issue above)

An alternative which could be faster in some rare circum stances, although unlikely not as I point out in the github issue above, would be to FOREGO the sync in all of these update functions , add a `dirty` flag to the struct overall , which could be read directly from the underlying record (htslib uses a dirty flag) ,a nd then check for dirty in any READ methods, and sync if necessary.

But probably it's just easier to add a sync here since we're alreadying syncing elsewhere in other writer fns

charlesgregory replied:
Yeah I tried to add that anywhere I was altering a header line or adding a new one. Will add.

charlesgregory replied:
addressed in f2b29df

jblachly commented on source/dhtslib/vcf/header.d:
```diff
@@ -61,4 +318,415 @@ struct VCFHeader
+
+///
+debug(dhtslib_unittest)
+unittest
```
Since no VCF file is actually used, this is a great unit test

charlesgregory replied:
Yeah I need to go through the library and convert many of our "integration" tests that rely on the htslib test files to true unit tests.

jblachly replied:
Will leave for another day

jblachly reviewed (changes requested):
Everything needs to be run through `dfmt` also

charlesgregory commented on source/dhtslib/vcf/package.d:
```diff
@@ -3,4 +3,99 @@ module dhtslib.vcf;
+static immutable  HeaderLengthsStrings = ["FIXED",".","A","G","R"];
+
+/// Replacement for htslib BCF_DT_*
+enum HeaderDictTypes
```
My goal was to make the `BCF_*` enums names more implicitly informative. Though I realize I could do a better job naming some of the individual enum names (like `HeaderLengths.A`). I also now realize I should be explicit (and be compatible with htslib) and should have my enum values be assigned directly from said htslib enums.

charlesgregory replied:
addressed in d23bb12.

charlesgregory commented on source/dhtslib/vcf/reader.d:
```diff
@@ -160,39 +162,43 @@ struct VCFReaderImpl(CoordSystem cs, bool isTabixFile)
     /// InputRange interface: iterate over all records
     @property bool empty()
     {
-        //     int bcf_read(htsFile *fp, const(bcf_hdr_t) *h, bcf1_t *v);
```
Overall the code is very similar. Just in order to be follow phobos rules, the range `empty` method cannot move the range forward only `popFront` can. I have solved this in our other ranges just as I have here, moved the "iterator next" call to `popFront` and introduced an `initialized` variable that is set when you check if the range is empty. This allows your to copy the range and still work as intended and allow other ranges to copy it as well. The issue in the first place was that it would hinder the ability of say `map` or `filter` or other phobos range functions to correctly use the range, as they expect to call `empty` without moving the range forward. I had already fixed this issue with the `SAMReader` `RecordRange` as I ran into it several months ago.

jblachly replied:
As long as the range is extensively tested and passes -- including things like reading from an already expended range, etc. then I am good with this

charlesgregory commented on source/dhtslib/vcf/writer.d:
```diff
@@ -149,105 +130,42 @@ struct VCFWriter
-        string line = format("##FILTER=<ID=%s,Description=\"%s\">\0", id, description);
-        bcf_hdr_append(this.vcfhdr.hdr, line.ptr);
-    }
+    // deprecated("use VCFHeader methods instead")
```
I had intended on leaving behind `deprecated` warnings for all functions that had been moved from `VCFReader` to `VCFHeader` but the function signature had ended up being changes for several of them so I had removed some of the functions. Honestly, I probably lost track of it across all my changes.

charlesgregory replied:
addressed in
1056615

charlesgregory commented on source/dhtslib/vcf/package.d:
```diff
@@ -3,4 +3,99 @@ module dhtslib.vcf;
+/// Replacement for htslib BCF_UN_*
+enum UNPACK
+{
+    ALT =   1, // up to ALT inclusive
```
Very possible. Though I do think my enum name is clearer about the intent of what to unpack.

charlesgregory commented on source/dhtslib/vcf/record.d:
```diff
@@ -459,8 +711,12 @@ struct VCFRecord
+        
+        if(f == "") hts_log_warning(__FUNCTION__, "filter was empty");
+        const int fid = bcf_hdr_id2int(this.vcfheader.hdr, BCF_DT_ID, toStringz(f));
+        if (fid == -1) hts_log_warning(__FUNCTION__, format("filter not found in header (ignoring): %s", f) );
```
Probably, I tried not to alter previous methods too much. This was an older method.

jblachly replied:
It looks like setting a filter of -1 might cause undefined behavior since they never check for this. I would make the function return early if the filter was not found (i.e. after the warning above)

```C
int bcf_add_filter(const bcf_hdr_t *hdr, bcf1_t *line, int flt_id)
{
    if ( !(line->unpacked & BCF_UN_FLT) ) bcf_unpack(line, BCF_UN_FLT);
    int i;
    for (i=0; i<line->d.n_flt; i++)
        if ( flt_id==line->d.flt[i] ) break;
    if ( i<line->d.n_flt ) return 0;    // this filter is already set
    line->d.shared_dirty |= BCF1_DIRTY_FLT;
    if ( flt_id==0 )    // set to PASS
        line->d.n_flt = 1;
    else if ( line->d.n_flt==1 && line->d.flt[0]==0 )
        line->d.n_flt = 1;
    else
        line->d.n_flt++;
    hts_expand(int, line->d.n_flt, line->d.m_flt, line->d.flt);
    line->d.flt[line->d.n_flt-1] = flt_id;
    return 1;
}
```

charlesgregory replied:
addressed in
0910089

charlesgregory commented on source/dhtslib/vcf/record.d:
```diff
@@ -459,8 +711,12 @@ struct VCFRecord
-        return bcf_add_filter(this.vcfheader.hdr, this.line, 
-            bcf_hdr_id2int(this.vcfheader.hdr, BCF_DT_ID, toStringz(f)));
+        
+        if(f == "") hts_log_warning(__FUNCTION__, "filter was empty");
```
Probably, I tried not to alter previous methods too much. This was an older method. I will look into it.

jblachly replied:
Confirmed* returns -1; since `f` is a D string, you could do the following:

```
if (fid == -1) hts_log_warning(__FUNCTION__, format("ignoring filter not found in header: %s", f.empty() ? "(empty)" : f));
```


*-> https://github.com/samtools/htslib/blob/238fe32d8c7aa05d3ac75d2249c61d8e268be58f/vcf.c#L3287-L3292

charlesgregory replied:
addressed in
0910089

charlesgregory commented on source/dhtslib/vcf/record.d:
```diff
@@ -523,19 +780,19 @@ struct VCFRecord
-    void addInfo(T)(string tag, T[] data)
-    if(!is(T==immutable(char)))             // otherwise string::immutable(char)[] will match the other template
+    void addInfo(T)(string tag, T data)
+    if(isIntegral!(ElementType!T) || isFloatingPoint!(ElementType!T))             // otherwise string::immutable(char)[] will match the other template
```
I was getting issues for `string` types matching the the `array` overload. Looking at it I think I needed to force `string` to use the `array` method and not the single value method.

charlesgregory commented on source/dhtslib/vcf/record.d:
```diff
@@ -14,18 +14,278 @@ import core.vararg;
 
+/// Struct to aid in conversion of VCF info data
+/// into D types
+struct InfoField
```
So the reason for copying is because of data ownership. And yes it can be done without a copy, the copy is an easy fix for some edge cases where you reference the info data but the `VCFRecord` is allowed to go out of scope or the `dtor` allowed to be called. In that situation, the info data is obviously gone and any references are invalid. An alternate to this implementation is similar to what I have done with `SAMRecord`, `Cigar`, and `TagValue`. All structs get reference counting, `VCFRecord` would have to check the reference counts of the `InfoFields` and `FormatFields` at destruction to make sure they didn't have references before the `bcf1_t` is destroyed in the `dtor`. I can certainly implement this.

jblachly replied:
I have no objection to VCFHeader, and VCFHeader records being copied, but for the VCFRecord itself (and INFO and FORMAT) there will be a huge perofrmance hit when say parsing Gnomad etc. Probably this s hould mirror SAMRecord

charlesgregory replied:
addressed in
e91d936

charlesgregory commented on source/dhtslib/vcf/record.d:
```diff
@@ -547,22 +804,24 @@ struct VCFRecord
-            ret = bcf_update_format_string(this.vcfheader.hdr, this.line, toStringz(tag), toStringz(data));
+        else static if(isSomeString!T){
+            auto strs = [data].map!(toUTFz!(char*)).array;
+            ret = bcf_update_format_string(this.vcfheader.hdr, this.line, toStringz(tag), strs.ptr, 1);
```
Yeah that is the only reason I didn't make the copy. I have tried to really think about preventing memory issues with this PR. I will double check all of this again before merge.

charlesgregory commented on source/dhtslib/vcf/package.d:
```diff
@@ -3,4 +3,99 @@ module dhtslib.vcf;
+}
+
+/// Strings for HeaderRecordType
+static immutable HeaderRecordTypeStrings = ["FILTER","INFO","FORMAT","contig", "struct", "generic"];
```
Unfortunately no. struct is a keyword.

charlesgregory replied:
Addressed in 9e8ea83.

charlesgregory commented on source/dhtslib/vcf/package.d:
```diff
@@ -3,4 +3,99 @@ module dhtslib.vcf;
+/// Replacement for htslib BCF_HL_*
+enum HeaderRecordType
+{
+    NULL = -1,
```
addressed in d23bb12.

charlesgregory commented on source/dhtslib/vcf/package.d:
```diff
@@ -3,4 +3,99 @@ module dhtslib.vcf;
+}
+
+/// Strings for HDR_LENGTH
+static immutable  HeaderLengthsStrings = ["FIXED",".","A","G","R"];
```
Unfortunately

charlesgregory commented on source/dhtslib/vcf/package.d:
```diff
@@ -3,4 +3,99 @@ module dhtslib.vcf;
+/// Replacement for htslib BCF_BT_*
+enum RecordType
+{
+    NULL =   0,  /// null
```
addressed in d23bb12.

charlesgregory commented on source/dhtslib/vcf/package.d:
```diff
@@ -3,4 +3,99 @@ module dhtslib.vcf;
+static immutable HeaderRecordTypeStrings = ["FILTER","INFO","FORMAT","contig", "struct", "generic"];
+
+/// Replacement for htslib BCF_HT_*
+enum HeaderTypes
```
Addressed in 9e8ea83.

charlesgregory commented on source/dhtslib/vcf/package.d:
```diff
@@ -3,4 +3,99 @@ module dhtslib.vcf;
+}
+
+/// Strings for HeaderTypes
+static immutable HeaderTypesStrings = ["Flag","Integer","Float","String","Character","Long"];
```
Addressed in 9e8ea83.

charlesgregory commented on source/dhtslib/vcf/package.d:
```diff
@@ -3,4 +3,99 @@ module dhtslib.vcf;
+
+/// Replacement for htslib BCF_VL_*
+enum HeaderLengths
+{
```
Addressed in 9e8ea83.

charlesgregory commented on source/dhtslib/vcf/package.d:
```diff
@@ -3,4 +3,99 @@ module dhtslib.vcf;
+}
+
+/// Replacement for htslib BCF_UN_*
+enum UNPACK
```
Second part resolved in 9e8ea83

charlesgregory commented:
Alright I added several changes to the enums. They now refer directly to the htslib `BCF_*` enums. I expanded some of the enum names to make them more self-explanatory. I made a function for calling `bcf_unpack` to standardize it for all functions along with unittests to confirm it works. 

I also changed the default `MAX_UNPACK` value to `UnpackLevels.None` as this should help performance by default and help us better catch bugs where we have not correctly unpacked data before using it. You may disagree with this sentiment. Let me know.

Still todo:
- Decide on no-copy vs copy for `InfoField` and `FormatField`
- Convert integration tests to actual unittests
- move integration tests to tests folder
- dfmt

jblachly commented on source/dhtslib/vcf/package.d:
```diff
@@ -3,4 +3,139 @@ module dhtslib.vcf;
+/// data in the bcf1_t. Values are inclusive 
+/// i.e UnpackLevels.AltAllele unpacks all data before and including the ALT allele
+/// Replacement for htslib BCF_UN_*
+enum UnpackLevels
```
Enum types are typically named in the singular (look at your others, `XXXType`). This makes them look more sensibile when typing function parameters

charlesgregory replied:
addressed in 
99f6949

jblachly reviewed (changes requested):
ALmost there. There were a few conversations not addressed (minor stuff) and a few you did comment on that I replied to.

The only major change I am requesting is no-copy VCFRecord, to match SAMRecord ; I could be persuaded either way on INFO and FORMAT structs

charlesgregory commented on source/dhtslib/vcf/record.d:
```diff
@@ -14,18 +14,278 @@ import core.vararg;
+
+/// Struct to aid in conversion of VCF format data
+/// into D types
+struct FormatField
```
addressed in
e91d936

charlesgregory commented on source/dhtslib/vcf/writer.d:
```diff
@@ -149,105 +130,42 @@ struct VCFWriter
-        string filter = format("##FILTER=<ID=%s,Description=\"%s\">\0", id, description);
-        bcf_hdr_append(this.vcfhdr.hdr, filter.ptr);
-    }
+    // deprecated("use VCFHeader methods instead")
```
addressed in
1056615

charlesgregory commented on source/dhtslib/vcf/writer.d:
```diff
@@ -149,105 +130,42 @@ struct VCFWriter
-
-        bcf_hdr_append(this.vcfhdr.hdr, line.ptr);
-    }
+    // deprecated("use VCFHeader methods instead")
```
addressed in
1056615

charlesgregory commented:
I think this should be ready to go.
